arch/arm/core/aarch64/cache.c

```diff
@@ -0,0 +1,59 @@
+/* cache.c - d-cache support for AARCH64 CPUs */
+
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief d-cache manipulation
+ *
+ * This module contains functions for manipulation of the d-cache.
+ */
+
+#include <cache.h>
+
+#define	CTR_EL0_DMINLINE_SHIFT	16
+#define	CTR_EL0_DMINLINE_MASK	GENMASK(19, 16)
+#define	CTR_EL0_CWG_SHIFT	24
+#define	CTR_EL0_CWG_MASK	GENMASK(27, 24)
+
+int arch_dcache_flush(void *addr, size_t size);
+int arch_dcache_invd(void *addr, size_t size);
+
+static size_t dcache_line_size;
+
+int arch_dcache_range(void *addr, size_t size, int op)
+{
+	if (op == K_CACHE_INVD) {
+		arch_dcache_invd(addr, size);
+	} else if (op == K_CACHE_WB_INVD) {
+		arch_dcache_flush(addr, size);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+size_t arch_dcache_line_size_get(void)
+{
+	uint64_t ctr_el0;
+	uint32_t cwg;
+	uint32_t dminline;
+
+	if (dcache_line_size)
+		return dcache_line_size;
```

Comment:
I understand the reasoning behind caching this value; this is somewhat dangerous in general. In particular:

1. There are big.LITTLE setups where different cores have different cache line sizes (see the amusing blog post https://www.mono-project.com/news/2016/09/12/arm64-icache/)
2. A sufficiently pessimistic compiler could turn this into incorrect code. In particular given the combination of "compilers assume that data races don't exist" and "this code is inherently race-condition prone".

E.g. if the compiler decided to "optimize" this code into, roughly:

```c
size_t *const dcache_line_size_ptr= &dcache_line_size;
size_t temp_a = *dcache_line_size_ptr;
size_t temp_b = *dcache_line_size_ptr;
if (!temp_b) {
    const size_t local_dcache_line_size = <snip rest of code>
    *dcache_line_size_ptr = local_dcache_line_size;
    temp_a = local_dcache_line_size;
}
return temp_a;
```
(with implicit memory accesses made explicit here.)

...which so far as I can tell is allowed (as it's the same behavior in singlethreaded execution - although it'd have to be a rather suboptimal compiler - loading `dcache_line_size_ptr` multiple times like this is almost never worth it on ARM).

```
AArch64 Fail_Cache
{
0:X0=val;
1:X0=val;
}
 P0             | P1            ;       
 MOV X2, #1     | MOV X2, #1    ;
 LDR X3, [X0]   | LDR X3, [X0]  ;
 LDR X1, [X0]   | LDR X1, [X0]  ;
 CBNZ X1, skipA | CBNZ X1, skipB; 
 STR X2, [X0]   | STR X2, [X0]  ;
 MOV X3, X2     | MOV X3, X2    ;
 skipA:         | skipB:        ;
exists
(1:X3=0)
```
(herd7 model - http://diy.inria.fr/www/index.html?record=aarch64)

Means that if 2 (or more) cores called this function concurrently then one of them could return `0` as the cache line size (!).

Reply:
You are correct. But I am not sure whether there is requirement to make zephyr runs on big/little systems, I have no such requirement. If your systems requires zephyr runs on big/little systems, would you submit patches based on this PR?

Reply:
> I have no such requirement.

I have no such requirement currently either - this is mostly an illustration of the issues with caching a value like this in an SMP system.